onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -1431,15 +1429,17 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<Node*>& fuse
                 shapes_opt[j] = tensor_shape_value;
                 shapes_max[j] = tensor_shape_value;
               }
-              dimension_update[input_name] = true;
+              engine_update = true;
             }
 
-            if (trt_profile == nullptr) {
-              trt_profile = trt_builder->createOptimizationProfile();
+            if (engine_update) {
```

Comment:
If put engine_update as a condition here, all dynamic shape tensors without changes on their shapes in current run will not call setShapeValues() and that will defeat the bug fix purpose of https://github.com/microsoft/onnxruntime/pull/5229.
Is the memory leak in your case due to setShapeValues()?

Reply:
We set engine_update to true if _any_ of the dynamic shape tensors has a change on its shape. The leak is due to calling `createOptimizationProfile()` every run. We cannot call setShapeValues() if trt_profile is nullptr. However, and please tell me if I'm wrong, the way we set engine_update to true will solve #5229 as well, correct?

Reply:
The original code checks trt_profile first. If it's nullptr, it will be created and then setDimensions() will be called. Not sure how the leak will happen.
for https://github.com/microsoft/onnxruntime/pull/5229, let's say a subgraph has 2 dynamic shape inputs, A and B. If at one inference run, only B's shape changed, per Nvidia's API requirement, trt_profile should be set by setDimensions() for both A and B, not just B. If you put engine_update as a condition, for the case B is coming after A, trt_profile will be set for B only. That's the issue https://github.com/microsoft/onnxruntime/pull/5229 to solve.

Reply:
Ok, then I think we should create one optimization profile in `TensorrtExecutionProvider::Compile()` and reuse it. The documentation says not to free the returned raw pointer, but we keep allocating new profile every run. This is where the memory is leaking.

Reply:
I implemented the new approach. I passed trt_profile as a part of trt_state. Please check this solution and tell me if it works for you.